### PR TITLE
WIP: GDPR Compliance

### PR DESCRIPTION
**Pull request by gymad**
## Description

Three new options in Administration / Campaign Email Settings page:

- "Opt In Checkbox on Web-To-Person form enabled": Contacts/Leads/Targets able to opt-in via Web-To-Person form directly.
- "Opt In Confirmation Email enabled": Contacts/Leads/Targets will get an Opt-In Confirmation email after submit the Web-To-Person form.
- "Opt-in confirmation email template": Template of Opt-In Confirmation Email, use $emailaddress_opt_in_link placeholder for Opt-In link.

- image:
_optincampaignemailsettings

- image:
_optinconfirmationlink

Campaing / Create Person Form: Opt in link automatically generated if "Opt In Checkbox on Web-To-Person form enabled" checkbox enabled on Administration / Campaign Email Settings page.
- image:
_optincheckboxonwebform

Opt-in link tracker on Campaign wizard / Template page for old user Opt-In marketing
-image:
_optintrackerlink

Email address field shows Opt In status on Detail View:
- image:
image

Email address fields shows Opt In status on Edit View:
- image:
image

## How To Test This

Users should able to Opt In view Web-To-Person forms and Opt-In Confirmation email, 
Campaign will sending marketing email only if:
- User Opted In
- User Not Opted Out
- (or) Marketing email contains an Opt-In tracker link

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

**Comment by pgorod** [comment]
@gymad how mature is this for testing? I have a friend who's very interested in this feature and I was hoping he could help test, but I don't know if it's too soon to start, perhaps you are still adding commits?

**Comment by gymad** [comment]
Hi @pgorod, thanks for your comment noticed me to change the title of this PR and I will add more details to description soon.

File: metadata/email_addressesMetaData.php [hunk 1]
```diff
@@ -34,89 +34,124 @@
...
+            'name' => 'opt_out',
+            'type' => 'bool',
+            'default' => 0,
+            'vname' => 'LBL_OPT_OUT',
+        ),
+        'opt_in_link' => array(
+            'name' => 'opt_in_link',
+            'vname' => 'LBL_OPT_IN_LINK',
+            'type' => 'function',
+//            'function_class' => 'OptInLink',
```
Comment (daniel-samson) [inline metadata/email_addressesMetaData.php]: Please remove  instead of commenting out.

File: modules/EmailMan/EmailMan.php [hunk 2]
```diff
@@ -413,553 +463,671 @@ function create_ref_email($marketing_id,$subject,$body_text,$body_html,$campagin
...
-		global $mod_strings;
+    /**
+     * @param $request
+     * @return bool
+     */
+    private function hasEmailOptInLink($request)
+    {
+        //fetch email marketing.
+        if (empty($this->current_emailmarketing) or !isset($this->current_emailmarketing)) {
+            if (!class_exists('EmailMarketing')) {
```
Comment (daniel-samson) [inline modules/EmailMan/EmailMan.php]: empty if statement? What is this meant for?

**Pull request closed without merging**